File: src/scripting/data_flow_analysis/ControlFlowGraph.ts
import { Instruction, InstructionSegment, Segment, SegmentType } from "../instructions";
import { Opcode } from "../opcodes";

export enum BranchType {
    None,
    Return,
    Jump,
    ConditionalJump,
    Call,
}

export class BasicBlock {
    readonly from: BasicBlock[] = [];
    readonly to: BasicBlock[] = [];

    constructor(
        readonly segment: InstructionSegment,
        readonly start: number,
        readonly end: number,
        readonly branch_type: BranchType,
        /**
         * Either jumps or calls, depending on `branch_type`.
         */
        readonly branch_labels: number[]
    ) {}

    link_to(other: BasicBlock): void {
        if (!this.to.includes(other)) {
            this.to.push(other);
            other.from.push(this);
        }
    }
}

export class ControlFlowGraph {
    readonly blocks: BasicBlock[] = [];

    private readonly instructions_to_block: Map<Instruction, BasicBlock> = new Map();
    private readonly labels_to_block = new Map<number, BasicBlock>();

    get_block_for_instuction(instruction: Instruction): BasicBlock | undefined {
        return this.instructions_to_block.get(instruction);
    }

    get_block_for_label(label: number): BasicBlock | undefined {
        return this.labels_to_block.get(label);
    }

    static create(segments: InstructionSegment[]): ControlFlowGraph {
        const cfg = new ControlFlowGraph();
        // Mapping of labels to basic blocks.

        for (const segment of segments) {
            this.create_basic_blocks(cfg, segment);
        }

        this.link_blocks(cfg);
        return cfg;
    }

    private static create_basic_blocks(cfg: ControlFlowGraph, segment: InstructionSegment) {
        const len = segment.instructions.length;
        let start = 0;
        let first_block = true;

        for (let i = 0; i < len; i++) {
            const inst = segment.instructions[i];

            let branch_type: BranchType;
            let branch_labels: number[];

            switch (inst.opcode) {
                // Return.
                case Opcode.RET:
                    branch_type = BranchType.Return;
                    branch_labels = [];
                    break;

                // Unconditional jump.
                case Opcode.JMP:
                    branch_type = BranchType.Jump;
                    branch_labels = [inst.args[0].value];
                    break;

                // Conditional jumps.
                case Opcode.JMP_ON:
                case Opcode.JMP_OFF:
                    branch_type = BranchType.ConditionalJump;
                    branch_labels = [inst.args[0].value];
                    break;
                case Opcode.JMP_E:
                case Opcode.JMPI_E:
                case Opcode.JMP_NE:
                case Opcode.JMPI_NE:
                case Opcode.UJMP_G:
                case Opcode.UJMPI_G:
                case Opcode.JMP_G:
                case Opcode.JMPI_G:
                case Opcode.UJMP_L:
                case Opcode.UJMPI_L:
                case Opcode.JMP_L:
                case Opcode.JMPI_L:
                case Opcode.UJMP_GE:
                case Opcode.UJMPI_GE:
                case Opcode.JMP_GE:
                case Opcode.JMPI_GE:
                case Opcode.UJMP_LE:
                case Opcode.UJMPI_LE:
                case Opcode.JMP_LE:
                case Opcode.JMPI_LE:
                    branch_type = BranchType.ConditionalJump;
                    branch_labels = [inst.args[2].value];
                    break;
                case Opcode.SWITCH_JMP:
                    branch_type = BranchType.ConditionalJump;
                    branch_labels = inst.args.slice(1).map(a => a.value);
                    break;

                // Calls.
                case Opcode.CALL:
                    branch_type = BranchType.Call;
                    branch_labels = [inst.args[0].value];
                    break;
                case Opcode.VA_CALL:
                    branch_type = BranchType.Call;
                    branch_labels = [inst.args[0].value];
                    break;
                case Opcode.SWITCH_CALL:
                    branch_type = BranchType.Call;
                    branch_labels = inst.args.slice(1).map(a => a.value);
                    break;

                // All other opcodes.
                default:
                    if (i === len - 1) {
                        branch_type = BranchType.None;
                        branch_labels = [];
                        break;
                    } else {
                        continue;
                    }
            }

            const block = new BasicBlock(segment, start, i + 1, branch_type, branch_labels);

            for (let j = block.start; j < block.end; j++) {
                cfg.instructions_to_block.set(block.segment.instructions[j], block);
            }

            cfg.blocks.push(block);

            if (first_block) {
                for (const label of segment.labels) {
                    cfg.labels_to_block.set(label, block);
                }

                first_block = false;
            }

            start = i + 1;
        }
    }

    private static link_blocks(cfg: ControlFlowGraph): void {
        // Pairs of calling block and block to which callees should return to.
        const callers: [BasicBlock, BasicBlock][] = [];

        for (let i = 0; i < cfg.blocks.length; i++) {
            const block = cfg.blocks[i];
            const next_block = cfg.blocks[i + 1];

            switch (block.branch_type) {
                case BranchType.Return:
                    continue;
                case BranchType.Call:
                    if (next_block) {
                        callers.push([block, next_block]);
                    }
                    break;
                case BranchType.None:
                case BranchType.ConditionalJump:
                    if (next_block) {
                        block.link_to(next_block);
                    }
                    break;
            }

            for (const label of block.branch_labels) {
                const to_block = cfg.labels_to_block.get(label);

                if (to_block) {
                    block.link_to(to_block);
                }
            }
        }

        for (const [caller, ret] of callers) {
            link_returning_blocks(cfg.labels_to_block, ret, caller);
        }
    }
}

/**
 * Links returning blocks to their callers.
 *
 * @param ret Block the caller should return to.
 * @param caller Calling block.
 */
function link_returning_blocks(
    label_blocks: Map<number, BasicBlock>,
    ret: BasicBlock,
    caller: BasicBlock
): void {
    for (const label of caller.branch_labels) {
        const callee = label_blocks.get(label);

        if (callee) {
            if (callee.branch_type === BranchType.Return) {
                callee.link_to(ret);
            } else {
                link_returning_blocks_recurse(new Set(), ret, callee);
            }
        }
    }
}

/**
 * @param encountered For avoiding infinite loops.
 */
function link_returning_blocks_recurse(
    encountered: Set<BasicBlock>,
    ret: BasicBlock,
    block: BasicBlock
): void {
    if (encountered.has(block)) {
        return;
    } else {
        encountered.add(block);
    }

    for (const to_block of block.to) {
        if (to_block.branch_type === BranchType.Return) {
            to_block.link_to(ret);
        } else {
            link_returning_blocks_recurse(encountered, ret, to_block);
        }
    }
}

File: lib/src/commonMain/kotlin/world/phantasmal/lib/assembly/dataFlowAnalysis/ControlFlowGraph.kt
package world.phantasmal.lib.assembly.dataFlowAnalysis

import world.phantasmal.lib.assembly.*

// See https://en.wikipedia.org/wiki/Control-flow_graph.

enum class BranchType {
    None,
    Return,
    Jump,
    ConditionalJump,
    Call,
}

/**
 * Instruction sequence into which control flow only enters at the start and only leaves at the end.
 * No code jumps/returns/calls into the middle of a basic block or branches out of a basic block
 * from the middle.
 */
interface BasicBlock {
    /**
     * The instruction segment that this block is a part of.
     */
    val segment: InstructionSegment

    /**
     * Index of this block's first instruction.
     */
    val start: Int

    /**
     * Index of the instruction right after this block's last instruction.
     */
    val end: Int

    /**
     * The way control flow leaves this block.
     */
    val branchType: BranchType

    /**
     * Either jumps or calls when non-empty, depending on [branchType].
     */
    val branchLabels: List<Int>

    /**
     * The blocks which branch to this block.
     */
    val from: List<BasicBlock>

    /**
     * The blocks this block branches to.
     */
    val to: List<BasicBlock>

    fun indexOfInstruction(instruction: Instruction): Int
}

/**
 * Graph representing the flow of control through the [BasicBlock]s of a script.
 */
class ControlFlowGraph(
    val blocks: List<BasicBlock>,
    private val instructionToBlock: Map<Instruction, BasicBlock>,
) {
    fun getBlockForInstruction(instruction: Instruction): BasicBlock {
        val block = instructionToBlock[instruction]
        requireNotNull(block) { "Instruction is not part of the control-flow graph." }
        return block
    }

    companion object {
        fun create(segments: List<InstructionSegment>): ControlFlowGraph {
            val cfg = ControlFlowGraphBuilder()

            // Mapping of labels to basic blocks.
            for (segment in segments) {
                createBasicBlocks(cfg, segment)
            }

            linkBlocks(cfg)
            return cfg.build()
        }
    }
}

private class ControlFlowGraphBuilder {
    val blocks: MutableList<BasicBlockImpl> = mutableListOf()
    val instructionsToBlock: MutableMap<Instruction, BasicBlockImpl> = mutableMapOf()
    val labelsToBlock: MutableMap<Int, BasicBlockImpl> = mutableMapOf()

    fun build(): ControlFlowGraph =
        ControlFlowGraph(blocks, instructionsToBlock)
}

private class BasicBlockImpl(
    override val segment: InstructionSegment,
    override val start: Int,
    override val end: Int,
    override val branchType: BranchType,
    override val branchLabels: List<Int>,
) : BasicBlock {
    override val from: MutableList<BasicBlockImpl> = mutableListOf()
    override val to: MutableList<BasicBlockImpl> = mutableListOf()

    override fun indexOfInstruction(instruction: Instruction): Int {
        var index = -1

        for (i in start until end) {
            if (instruction == segment.instructions[i]) {
                index = i
                break
            }
        }

        return index
    }

    fun linkTo(other: BasicBlockImpl) {
        if (other !in to) {
            to.add(other)
            other.from.add(this)
        }
    }
}

private fun createBasicBlocks(cfg: ControlFlowGraphBuilder, segment: InstructionSegment) {
    val len = segment.instructions.size
    var start = 0
    var firstBlock = true

    for (i in 0 until len) {
        val inst = segment.instructions[i]

        var branchType: BranchType
        var branchLabels: List<Int>

        when (inst.opcode) {
            // Return.
            OP_RET -> {
                branchType = BranchType.Return
                branchLabels = emptyList()
            }

            // Unconditional jump.
            OP_JMP -> {
                branchType = BranchType.Jump
                branchLabels = listOf(inst.args[0].value as Int)
            }

            // Conditional jumps.
            OP_JMP_ON,
            OP_JMP_OFF,
            -> {
                branchType = BranchType.ConditionalJump
                branchLabels = listOf(inst.args[0].value as Int)
            }
            OP_JMP_E,
            OP_JMPI_E,
            OP_JMP_NE,
            OP_JMPI_NE,
            OP_UJMP_G,
            OP_UJMPI_G,
            OP_JMP_G,
            OP_JMPI_G,
            OP_UJMP_L,
            OP_UJMPI_L,
            OP_JMP_L,
            OP_JMPI_L,
            OP_UJMP_GE,
            OP_UJMPI_GE,
            OP_JMP_GE,
            OP_JMPI_GE,
            OP_UJMP_LE,
            OP_UJMPI_LE,
            OP_JMP_LE,
            OP_JMPI_LE,
            -> {
                branchType = BranchType.ConditionalJump
                branchLabels = listOf(inst.args[2].value as Int)
            }
            OP_SWITCH_JMP -> {
                branchType = BranchType.ConditionalJump
                branchLabels = inst.args.drop(1).map { it.value as Int }
            }

            // Calls.
            OP_CALL,
            OP_VA_CALL,
            -> {
                branchType = BranchType.Call
                branchLabels = listOf(inst.args[0].value as Int)
            }
            OP_SWITCH_CALL -> {
                branchType = BranchType.Call
                branchLabels = inst.args.drop(1).map { it.value as Int }
            }

            // All other opcodes.
            else -> {
                if (i == len - 1) {
                    // This is the last block of the segment.
                    branchType = BranchType.None
                    branchLabels = emptyList()
                } else {
                    // Non-branching instruction, part of the current block.
                    continue
                }
            }
        }

        val block = BasicBlockImpl(segment, start, i + 1, branchType, branchLabels)

        for (j in block.start until block.end) {
            cfg.instructionsToBlock[block.segment.instructions[j]] = block
        }

        cfg.blocks.add(block)

        if (firstBlock) {
            for (label in segment.labels) {
                cfg.labelsToBlock[label] = block
            }

            firstBlock = false
        }

        start = i + 1
    }
}

private fun linkBlocks(cfg: ControlFlowGraphBuilder) {
    // Pairs of calling block and block to which callees should return to.
    val callers = mutableListOf<Pair<BasicBlockImpl, BasicBlockImpl>>()

    for (i in cfg.blocks.indices) {
        val block = cfg.blocks[i]
        val nextBlock = cfg.blocks.getOrNull(i + 1)

        when (block.branchType) {
            BranchType.Return ->
                continue

            BranchType.Call ->
                nextBlock?.let { callers.add(block to nextBlock); }

            BranchType.None,
            BranchType.ConditionalJump,
            -> nextBlock?.let(block::linkTo)

            else -> {
                // Ignore.
            }
        }

        for (label in block.branchLabels) {
            cfg.labelsToBlock[label]?.let { toBlock ->
                block.linkTo(toBlock)
            }
        }
    }

    for ((caller, ret) in callers) {
        linkReturningBlocks(cfg.labelsToBlock, ret, caller)
    }
}

/**
 * Links returning blocks to their callers.
 *
 * @param labelBlocks Mapping of labels to basic blocks.
 * @param ret Basic block the caller should return to.
 * @param caller Calling basic block.
 */
private fun linkReturningBlocks(
    labelBlocks: Map<Int, BasicBlockImpl>,
    ret: BasicBlockImpl,
    caller: BasicBlockImpl,
) {
    for (label in caller.branchLabels) {
        labelBlocks[label]?.let { callee ->
            if (callee.branchType === BranchType.Return) {
                callee.linkTo(ret)
            } else {
                linkReturningBlocksRecurse(mutableSetOf(), ret, callee)
            }
        }
    }
}

/**
 * @param encountered For avoiding infinite loops.
 * @param ret
 * @param block
 */
private fun linkReturningBlocksRecurse(
    encountered: MutableSet<BasicBlockImpl>,
    ret: BasicBlockImpl,
    block: BasicBlockImpl,
) {
    if (block in encountered) {
        return
    } else {
        encountered.add(block)
    }

    for (toBlock in block.to) {
        if (toBlock.branchType === BranchType.Return) {
            toBlock.linkTo(ret)
        } else {
            linkReturningBlocksRecurse(encountered, ret, toBlock)
        }
    }
}
